Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Smoothing/curvature_flow_impl.h

```diff
@@ -0,0 +1,303 @@
+      else
+      {
+        A.set_coef(t.get<0>(), t.get<1>(), diagonal_[t.get<0>()] -time * t.get<2>(), true);
+      }
+    }
+
+    apply_constraints(A);
+    // we do not call A.assemble_matrix here
+    // Eigen's compute during factorization does the building correctly,
+    // and without assemble_matrix the reference A can be used in the next iterations.
+  }
+
+  void calculate_D_diagonal(std::vector<double>& diagonal)
+  {
+    // calculates all elements on the diagonal of D
+    // and marks the entries that correspond to constrained vertices
+
+    diagonal.assign(nb_vert_, 0);
+    constrained_flags_.assign(nb_vert_, false);
+
+    BOOST_FOREACH(face_descriptor f, frange_)
+    {
+      double area = face_area(f, mesh_);
+      BOOST_FOREACH(vertex_descriptor v, vertices_around_face(halfedge(f, mesh_), mesh_))
+      {
+        int idx = vimap_[v];
+        if(!is_constrained(v))
+        {
+          diagonal[idx] += area / 6.0; // 2*area/12
+        }
+        else
+        {
+          constrained_flags_[idx] = true;
+          diagonal[idx] = 1.0; // needed at multiplication D*b when computing rhs
+        }
+      }
+    }
+  }
+
+  void compute_rhs(Eigen_vector& bx, Eigen_vector& by, Eigen_vector& bz)
+  {
+    CGAL_assertion(diagonal_.size() == nb_vert_);
+    CGAL_assertion(bx.size() == nb_vert_);
+    CGAL_assertion(by.size() == nb_vert_);
+    CGAL_assertion(bz.size() == nb_vert_);
+
+    BOOST_FOREACH(vertex_descriptor vi, vrange_)
+    {
+      int index = vimap_[vi];
+      Point p = get(vpmap_, vi);
+      bx.set(index, p.x());
+      by.set(index, p.y());
+      bz.set(index, p.z());
+    }
+
+    // multiply D * b
+    for(int i = 0; i < nb_vert_; ++i)
+    {
+      bx[i] *= diagonal_[i];
+      by[i] *= diagonal_[i];
+      bz[i] *= diagonal_[i];
+    }
+  }
+
+  // handling constrains
+  // -----------------------------------
+  void apply_constraints(Eigen_matrix& A)
+  {
+    CGAL_assertion(constrained_flags_.size() == nb_vert_);
+    for(int i = 0; i < nb_vert_; ++i)
+    {
+      if(constrained_flags_[i] == true)
+        A.set_coef(i, i, 1.0, true);
+    }
+  }
+
+  bool is_constrained(const vertex_descriptor& v)
+  {
+    return get(vcmap_, v);
+  }
+
+  template<typename FaceRange>
+  void check_face_range(const FaceRange& face_range)
+  {
+    BOOST_FOREACH(face_descriptor f, face_range)
+    {
+      frange_.insert(f);
+      BOOST_FOREACH(vertex_descriptor v, vertices_around_face(halfedge(f, mesh_), mesh_))
+        vrange_.insert(v);
+    }
+  }
+
+  private:
+  // data members
+  // ------------
+  std::size_t nb_vert_;
+  PolygonMesh& mesh_;
+  VertexPointMap& vpmap_;
+  VertexConstraintMap& vcmap_;
+  IndexMap vimap_;
+
+  // linear system data
+  std::vector<double> diagonal_; // index of vector -> index of vimap_
+  std::vector<bool> constrained_flags_;
+
+  std::set<face_descriptor> frange_;
+  std::set<vertex_descriptor> vrange_;
```

Comment:
hang on, I am confused: do you mean all vertices of the mesh?
I want only the vertices in that face range.

Reply:
no only the vertices of the selection

Reply:
AFAIU, we would use std::sort and std::unique on a std::vector so as to avoid the double loop in order to take out duplicates, but can we avoid the double loop when we deal with a subset of the faces of the mesh?

Reply:
when you collect the vertices per faces, collect them all, duplicates included. Then sort, call unique and remove the extra copies from the range. Then you're done. No idea if it does worth it in your case.